async_adapter.go

```diff
@@ -0,0 +1,261 @@
+		return err
+	}
+	a.msgCh <- choiceMsg{id}
+	return nil
+}
+
+// Abort stops the VM with the given error as soon as possible (either within
+// the current event, or on the next event). If a nil error is passed, Abort
+// will replace it with Stop (so that NodeComplete and DialogueComplete still
+// fire).
+func (a *AsyncAdapter) Abort(err error) error {
+	if old := a.state.Swap(VMStateStopped); old == VMStateStopped {
+		return ErrAlreadyStopped
+	}
+	if err == nil {
+		err = Stop
+	}
+	a.msgCh <- abortMsg{err}
+	return nil
+}
+
+// waitForGo waits for Go or Abort to be called.
+func (a *AsyncAdapter) waitForGo() error {
+	switch msg := (<-a.msgCh).(type) {
+	case goMsg:
+		return nil
+	case choiceMsg:
+		// This is incredibly unlikely, but I check it anyway.
+		return errors.New("AsyncAdapter.GoWithChoice called, but last event was not Options")
+	case abortMsg:
+		return msg.err
+	default:
+		return fmt.Errorf("invalid message type %T received", msg)
+	}
+}
+
+// waitForChoice waits for GoWithChoice or Abort to be called.
+func (a *AsyncAdapter) waitForChoice() (int, error) {
+	switch msg := (<-a.msgCh).(type) {
+	case goMsg:
+		// This is incredibly unlikely, but I check it anyway.
+		return -1, errors.New("AsyncAdapter.Go called, but last event was Options")
+	case choiceMsg:
+		return msg.choice, nil
+	case abortMsg:
+		return -1, msg.err
+	default:
+		return -1, fmt.Errorf("invalid message type %T received", msg)
+	}
+}
+
+// --- DialogueHandler implementation --- \\
+
+// NodeStart is called by the VM and passed through to the
+// AsyncDialogueHandler directly.
+func (a *AsyncAdapter) NodeStart(nodeName string) error {
+	return a.handler.NodeStart(nodeName)
+}
+
+// PrepareForLines is called by the VM and passed through to the
+// AsyncDialogueHandler directly.
+func (a *AsyncAdapter) PrepareForLines(lineIDs []string) error {
+	return a.handler.PrepareForLines(lineIDs)
+}
+
+// Line is called by the VM and blocks until Go or Abort is called.
+func (a *AsyncAdapter) Line(line Line) error {
+	if err := a.stateTransition(VMStateRunning, VMStatePaused); err != nil {
+		return err
+	}
+	a.handler.Line(line)
+	return a.waitForGo()
+}
+
+// Options is called by the VM and blocks until GoWithChoice or Abort is called.
+func (a *AsyncAdapter) Options(options []Option) (int, error) {
+	if err := a.stateTransition(VMStateRunning, VMStatePausedOptions); err != nil {
+		return -1, err
+	}
+	a.handler.Options(options)
+	return a.waitForChoice()
+}
+
+// Command is called by the VM and blocks until Go or Abort is called.
+func (a *AsyncAdapter) Command(command string) error {
+	if err := a.stateTransition(VMStateRunning, VMStatePaused); err != nil {
+		return err
+	}
+	a.handler.Command(command)
+	return a.waitForGo()
+}
+
+// NodeComplete is called by the VM and passed through to the
+// AsyncDialogueHandler directly.
+func (a *AsyncAdapter) NodeComplete(nodeName string) error {
+	return a.handler.NodeComplete(nodeName)
```

Comment:
**question:** Clearly you've added blocking for the three most useful places. Why not also provide the option to block during NodeComplete or DialogueComplete? I could see that potentially being useful perhaps for playing animations during the end of dialogue.

It can also be handled outside this lib by the caller, so this is just a question.

Reply:
I definitely agree with you about NodeComplete. DialogueComplete feels less clear since it's at the end, but I guess it might be desirable to block before returning from VirtualMachine.Run in order to delay whatever the user has put afterwards. 

From a consistency perspective, maybe all the methods should do the same thing? WDYT?

Reply:
That does make sense. Either they should all be able to do the same thing or a user should be able to pass in a config telling this lib which methods they would like to have block... I'm not sure whether that's an easy requirement to achieve or not. :T